compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp

```diff
@@ -189,74 +188,119 @@ resolveWorkgroupForAll(RewriterBase &rewriter, scf::ForallOp forallOp,
   if (failed(workgroupMapping)) {
     return failure();
   }
+  if (workgroupMapping->size() != mixedLowerBound.size()) {
+    return forallOp.emitOpError(
+        "expected as many workgroup mapping attributes as number of loops");
+  }
 
   OpBuilder::InsertionGuard g(rewriter);
   rewriter.setInsertionPoint(forallOp);
+  Location loc = forallOp.getLoc();
 
-  SmallVector<OpFoldResult> procId;
+  // Get process IDs and counts by querying hal.interface.workgroup.id/count ops
+  // and delinearizing any dimensions of the forall beyond `deLinearizeFrom`.
+  SmallVector<OpFoldResult> procIds, nProcs;
+  std::tie(procIds, nProcs) = getProcIdsAndNprocs(
+      forallOp, rewriter, loc, workgroupMapping.value(), mixedLowerBound,
+      mixedUpperBound, mixedStep, deLinearizeFrom);
 
-  {
-    FailureOr<std::pair<SmallVector<OpFoldResult>, SmallVector<OpFoldResult>>>
-        procInfo =
-            getProcIdsAndNprocs(forallOp, rewriter, forallOp.getLoc(),
-                                workgroupMapping.value(), mixedLowerBound,
-                                mixedUpperBound, mixedStep, deLinearizeFrom);
-    if (failed(procInfo)) {
-      return failure();
-    }
-    std::swap(procId, procInfo->first);
+  // Scale the process IDs and counts to account for the forall op steps. These
+  // are the forall offsets for each process, and the bounds of these offsets.
+  rewriter.setInsertionPoint(forallOp);
```

Comment:
Do we need to reset the insertion point?

Reply:
Agree, setting up an insertion guard in/around `getProcIdsAndNprocs` would be more standard.